services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java

```diff
@@ -166,13 +168,21 @@ public Response logout(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_PARAM) String
             return ErrorPage.error(session, null, Response.Status.BAD_REQUEST, Messages.INVALID_PARAMETER, OIDCLoginProtocol.REDIRECT_URI_PARAM);
         }
 
-        if (postLogoutRedirectUri != null && encodedIdToken == null) {
+        if (postLogoutRedirectUri != null && encodedIdToken == null && clientId == null) {
             event.event(EventType.LOGOUT);
             event.error(Errors.INVALID_REQUEST);
-            logger.warnf("Parameter 'id_token_hint' is required when 'post_logout_redirect_uri' is used.");
+            logger.warnf("Either the parameter 'client_id' or the parameter 'id_token_hint' is required when 'post_logout_redirect_uri' is used.");
             return ErrorPage.error(session, null, Response.Status.BAD_REQUEST, Messages.MISSING_PARAMETER, OIDCLoginProtocol.ID_TOKEN_HINT);
         }
 
+        boolean confirmationNeeded = true;
+        boolean forcedConfirmation = false;
+        ClientModel client = clientId == null ? null : realm.getClientByClientId(clientId);
+        if (clientId != null && client == null) {
```

Comment:
Any reason why don't fail fast here?

Reply:
Good point.

The reason is the potential client enumeration. There was recently some issue reported by customers related to client enumeration (discussed also recently on keycloak-security). In relation to that I've created 2 issues:
https://github.com/keycloak/keycloak/issues/12162 - Incorrect error messages used during failed OIDC client authentication
https://github.com/keycloak/keycloak/issues/12164 - Possible client enumeration in the authorization endpoint

There are probably more places and I am not sure how much we really care of client enumeration and if it is possible to avoid it at all. But at least, I wanted to avoid introducing another obvious place for it. So currently with unknown client_id there is guaranteed that confirmation screen will be displayed to the user. Which I personally consider OK from the security perspective. WDYT?